### DISCUSSION TITLE
ABP Issue

### DISCUSSION THREAD

**Discussion by Ahsan-Averos** [Lora-net/LoRaMac-node#1076] [Q&A]
can someone helping me out why is not confirming my data::
<debug> app: Uplink counter   : 120
<debug> app: Uplink Data rate : DR_0
<debug> app: Uplink TX power  : 0
<debug> app: Uplink frequency: 868500000
<debug> app: FDS update success! FileId: 0x3E8 RecKey:0x1
<debug> app: FDS update success! FileId: 0x3E8 RecKey:0x2
<debug> app: FDS update success! FileId: 0x3E8 RecKey:0x5
<info> app: Data stored on NVM: 224 bytes
<info> app: **MCPS-Request: MCPS_UNCONFIRMED,  OK**
This above line is MCPS-Request: MCPS_UNCONFIRMED,  OK, I can not find solution for this  Thanks

**Comment by mluis1** [comment]
What is the relation between your described problem and ABP activation? 
What do you mean by not confirming your data?

Issuing `MCPS-Request: MCPS_UNCONFIRMED` will never request the network server to acknowledge the transmitted data.

In case you want to receive an acknowledge from the network server you need to execute an `MCPS-Request: MCPS_CONFIRMED` command instead.

Please note that one should avoid using confirmed uplinks as much as possible. This kind of uplinks should only be used for example to report and validate an alarm situation.

Back to your issue I would recommend the reading of the LoRaWAN specification as well as to have a look at the provided examples.
In the provided examples you will see bellow function implementation which will execute a `MCPS_UNCONFIRMED` or `MCPS_CONFIRMED` uplink depending on the `isTxConfirmed` function parameter.

https://github.com/Lora-net/LoRaMac-node/blob/0d80d260d9b74e42892f07b6604e36fc4b1840cf/src/apps/LoRaMac/common/LmHandler/LmHandler.c#L469-L485

**Reply by Ahsan-Averos** [reply]
Sir do you have any sample code for NRF52 controller with  SX126x IC.
Thanks

**Reply by mluis1** [reply]
No, we don't.

The current supported platforms are listed on following `README.md` file chapter [Supported platforms](https://github.com/Lora-net/LoRaMac-node#supported-platforms)

I noticed that you have asked the same question on a project which looks to be a porting of this project for the NRF52 MCU.
https://github.com/insightsip/ISP4520-examples/issues/16

Although, your question is more related on how a LoRaWAN end-device is supposed to work. The best advice I can provide is to first read the LoRaWAN specifications and search the web for tutorials on implementing LoRaWAN end-devices.

Some readings you might find useful at below links:
LoRaWAN general: https://www.thethingsnetwork.org/docs/lorawan/ 
The devices section of: https://www.thethingsnetwork.org/docs/devices/
The gateways section of: https://www.thethingsnetwork.org/docs/gateways/
The network section of: https://www.thethingsnetwork.org/docs/network/
The applications and API sections of: https://www.thethingsnetwork.org/docs/applications/
